File: 3one_degree_freedom_system_response.m
clear all
clc

% System data
mass = 1;
damping_ratio = 0.003;   
equivalent_stiffness = 5e6;
no_damped_natural_frequency = sqrt(equivalent_stiffness/mass);
equivalent_viscous_damping = 2*mass*damping_ratio*no_damped_natural_frequency;
fn = no_damped_natural_frequency/(2*pi)            % no damped natural frequenequivalent viscous dampingy [Hz]

% Exequivalent viscous dampingitation data
A=0.7;
B=2.2;

% Time domain data
points = 2000;
delta_time = 0.00431;
final_time = delta_time*final_time;

equivalent_time = 0:delta_time:final_time;
rad_excitation_frequency = [30 no_damped_natural_frequency 3*no_damped_natural_frequency];

complex_amplitude = (A-1i*B)/2;
receptance = 1./(-mass*rad_excitation_frequency.^2+1i*equivalent_viscous_damping*rad_excitation_frequency+equivalent_stiffness);                                
response = receptance*complex_amplitude;
                                       
for it=1:length(equivalent_time)
    f1(it)=complex_amplitude*exp(1i*rad_excitation_frequency(1)*equivalent_time(it))+conj(complex_amplitude)*exp(-1i*rad_excitation_frequency(1)*equivalent_time(it));
    f2(it)=complex_amplitude*exp(1i*rad_excitation_frequency(2)*equivalent_time(it))+conj(complex_amplitude)*exp(-1i*rad_excitation_frequency(2)*equivalent_time(it));
    f3(it)=complex_amplitude*exp(1i*rad_excitation_frequency(3)*equivalent_time(it))+conj(complex_amplitude)*exp(-1i*rad_excitation_frequency(3)*equivalent_time(it));     
    x1(it)=response(1)*exp(1i*rad_excitation_frequency(1)*equivalent_time(it))+conj(response(1))*exp(-1i*rad_excitation_frequency(1)*equivalent_time(it));
    x2(it)=response(2)*exp(1i*rad_excitation_frequency(2)*equivalent_time(it))+conj(response(2))*exp(-1i*rad_excitation_frequency(2)*equivalent_time(it));
    x3(it)=response(3)*exp(1i*rad_excitation_frequency(3)*equivalent_time(it))+conj(response(3))*exp(-1i*rad_excitation_frequency(3)*equivalent_time(it));
end

% Results

figure(1)
subplot(2,3,1)
plot(equivalent_time,f1)
axis([0 0.5 -3 3])
ylabel('Excitation [EU]')

subplot(2,3,4)
plot(equivalent_time,x1)
axis([0 0.5 -5e-5 5e-5])
ylabel('Response [EU]')

subplot(2,3,2)
plot(equivalent_time,f2)
axis([0 0.5 -3 3])

subplot(2,3,5)
plot(equivalent_time,x2)
axis([0 0.5 -5e-5 5e-5])

subplot(2,3,3)
plot(equivalent_time,f3)
axis([0 0.5 -3 3])
xlabel('Time [s]')

subplot(2,3,6)
plot(equivalent_time,x3)
axis([0 0.5 -5e-5 5e-5])
xlabel('Time [s]')

% Response in the frequency domain for a given excitation frequency
inc=0.013;
frequency_array = 0:inc:1.2*max(rad_excitation_frequency);

receptance1 = 1./(-mass*frequency_array.^2+1i*equivalent_viscous_damping*frequency_array+equivalent_stiffness);

figure(2)
subplot(3,1,1)
plot(frequency_array/(2*pi),20*log10(abs(receptance1)))
ylabel('|Receptance|')
title(' ref=1 [m/N]')

force_array = zeros(1,length(frequency_array));
displacement_vector = zeros(1,length(frequency_array));

FFF=(A-1i*B)/2;    

% Excitation and response for each frequency                
omegaext=rad_excitation_frequency(1);                      
                                   
% Complex amplitude of the excitation
receptance2 = 1./(-mass*omegaext.^2+1i*equivalent_viscous_damping*omegaext+equivalent_stiffness);

response1 = receptance2*FFF;

for ii=1:length(frequency_array)
    if abs(omegaext-frequency_array(ii))<=inc 
           force_array(ii)=FFF;
           displacement_vector(ii)=response1; 
   end    
end


subplot(3,1,2)
plot(frequency_array/(2*pi),abs(force_array))
ylabel('|Excitation| [N]')

subplot(3,1,3)
plot(ome/(2*pi),abs(displacement_vector))               
ylabel('|Response| [m]')
xlabel('frequency [Hz]')

File: 3one_degree_freedom_system_response.py
import numpy 
import matplotlib

# Defining system data
mass = 1
damp_ratio = 0.003
stiff = 5*numpy.exp(6)
freq_rad = numpy.sqrt(stiff/mass)
damp = 2*mass*damp_ratio*freq_rad
natural_damp = freq_rad/(2*numpy.pi)

# Defining excitation data
a = 0.7
b = 2.2

excitation_freq = numpy.array([30, freq_rad, 3*freq_rad])
complex_amplitude = (a-1j*b)/2
receptance = 1. / (-mass * excitation_freq * numpy.exp(2) + 1j * excitation_freq * damp + stiff)
response = complex_amplitude*receptance

def get_time():
  step = 0.00431
  number_of_points = 2000
  final_time = step*number_of_points
  index = 0
  time_array = []
  while(index < number_of_points):
    if index == 0:
      time_array.append(0)
      index += 1
    else:
      time_array.append(time_array[index-1] + step)
      index += 1
  return numpy.array(time_array)

time = get_time()

func = [[],[],[]]
x = [[],[],[]]

for i in range(len(func)):
  for j in range(len(time)):
    x[i].append(response[i]*numpy.exp(1j*excitation_freq[i]*time[j])+numpy.conj(response[i])*numpy.exp(-1j*excitation_freq[i]*time[j]))
    func[i].append(complex_amplitude*numpy.exp(1j*excitation_freq[i]*time[j])+numpy.conj(complex_amplitude)*numpy.exp(-1j*excitation_freq[i]*time[j]))


figure, axis = matplotlib.pyplot.subplots(2)
figure.suptitle('F1 and X1')
axis[0].plot(time,func[0])
axis[0].axis([0, 0.5, -3, 3])
axis[0].set_xlabel('Time [s]')
axis[0].set_ylabel('Excitation [EU]')
axis[1].plot(time, x[0])
axis[1].axis([0, 0.5, -5*numpy.exp(-5), 5*numpy.exp(-5)])
axis[1].set_xlabel('Time [s]')
axis[1].set_ylabel('Response [EU]')

figure, axis = matplotlib.pyplot.subplots(2)
figure.suptitle('F2 and X2')
axis[0].plot(time,func[1])
axis[0].axis([0, 0.5, -3, 3])
axis[0].set_xlabel('Time [s]')
axis[0].set_ylabel('Excitation [EU]')
axis[1].plot(time, x[1])
axis[1].axis([0, 0.5, -5*numpy.exp(-5), 5*numpy.exp(-5)])
axis[1].set_xlabel('Time [s]')
axis[1].set_ylabel('Response [EU]')

figure, axis = matplotlib.pyplot.subplots(2)
figure.suptitle('F3 and X3')
axis[0].plot(time,func[2])
axis[0].axis([0, 0.5, -3, 3])
axis[0].set_xlabel('Time [s]')
axis[0].set_ylabel('Excitation [EU]')
axis[1].plot(time, x[2])
axis[1].axis([0, 0.5, -5*numpy.exp(-5), 5*numpy.exp(-5)])
axis[1].set_xlabel('Time [s]')
axis[1].set_ylabel('Response [EU]')

def get_vector():
  step = 0.013
  index = 0
  vector_freq = []
  while (index < 1.2*numpy.max(excitation_freq)):
    vector_freq.append(index+step)
    index += 1
  return numpy.array(vector_freq)

freq_vector = get_vector()
frf_receptance = 1. / -mass * freq_vector * numpy.exp(2) + 1j * freq_vector * damp + stiff

vector_force = numpy.zeroes(1, len(freq_vector))
displacement_vector = numpy.zeroes(1, len(freq_vector))
freq_response = (a-1j*b)/2

figure, axis = matplotlib.pyplot.subplots(3)
figure.suptitle('Receptance, Excitation and Response')
axis[0].plot(freq_vector/(2*numpy.pi), 20*numpy.log10(numpy.abs(frf_receptance)))
axis[0].axis([0, 0.5, -3, 3])
axis[0].set_title('ref=1 [m/N]')
axis[0].set_xlabel('Time [s]')
axis[0].set_ylabel('Receptance')
axis[1].plot(time, x[2])
axis[1].axis([0, 0.5, -5*numpy.exp(-5), 5*numpy.exp(-5)])
axis[1].set_xlabel('Time [s]')
axis[1].set_ylabel('Response [EU]')